core/templates/dev/head/pages/exploration_editor/editor_tab/StateSolutionHelperService.js

```diff
@@ -0,0 +1,117 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for solution editor.
+ */
+
+oppia.constant('SUPPORTED_INTERACTIONS', [
+  'InteractiveMap',
+  'MusicNotesInput',
+  'GraphInput',
+  'SetInput',
+  'MathExpressionInput',
+  'MultipleChoiceInput',
+  'ImageClickInput',
+  'ItemSelectionInput',
+  'LogicProof'
+]);
+
+oppia.constant('INTERACTION_OBJECT_TYPES', {
+  CodeRepl: 'CodeString',
+  GraphInput: 'Graph',
+  ImageClickInput: 'ImageWithRegions',
+  PencilCodeEditor: 'CodeString',
+  MathExpressionInput: 'UnicodeString',
+  MusicNotesInput: 'MusicPhrase',
+  NumericInput: 'Real',
+  TextInput: 'NormalizedString',
+  LogicProof: 'LogicQuestion'
+});
+
+oppia.factory('StateSolutionHelperService', [
+  'oppiaHtmlEscaper', 'responsesService', 'stateInteractionIdService',
+  'explorationContextService', 'editorContextService', '$injector',
+  'explorationStatesService', 'angularNameService', 'stateSolutionService',
+  'AnswerClassificationService', 'alertsService',
+  'SUPPORTED_INTERACTIONS', 'INTERACTION_OBJECT_TYPES',
+  function(oppiaHtmlEscaper, responsesService, stateInteractionIdService,
+    explorationStatesService, editorContextService, $injector,
+    explorationContextService, angularNameService, stateSolutionService,
+    AnswerClassificationService, alertsService,
+    SUPPORTED_INTERACTIONS, INTERACTION_OBJECT_TYPES) {
+    var currentSolutionIsValid = true;
+    return {
+      isSupportedInteraction: function (id) {
+        return (SUPPORTED_INTERACTIONS.indexOf(id) !== -1);
+      },
+      getInteractionObjectType: function (id) {
+        return INTERACTION_OBJECT_TYPES[id];
+      },
+      getCorrectAnswerObject: function(answer, objectType) {
+        if (objectType === 'CodeString') {
+          return {
+            code: answer,
+            output: '',
+            evaluation: '',
+            error: ''
+          }
+        } else {
+          return answer;
+        }
+      },
+      verifyAndSaveAnswer: function(explorationId, state, correctAnswer) {
+        var interactionId = stateInteractionIdService.savedMemento;
+        var rulesServiceName = (
+          angularNameService.getNameOfInteractionRulesService(interactionId));
+        // Inject RulesService dynamically.
+        var rulesService = $injector.get(rulesServiceName);
+        AnswerClassificationService.getMatchingClassificationResult(
+          explorationId, state, correctAnswer, true, rulesService
+        ).then(function(result) {
+          if (editorContextService.getActiveStateName() !== (
+              result.outcome.dest)) {
+            currentSolutionIsValid = true;
+            stateSolutionService.saveDisplayedValue();
+          } else {
+            alertsService.addInfoMessage('That solution does not lead ' +
```

Comment:
This helper service is not the place to do alerts or otherwise modify the frontend or other services. It should just, statelessly, validate a solution. Callers should take care of what happens onSuccess() or onFailure().